// yaml-import/dist-node/bin/main.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = main;

var _path = _interopRequireDefault(require("path"));

var _cliBelt = require("cli-belt");

var _commonTags = require("common-tags");

var _arg = _interopRequireDefault(require("arg"));

var _jsYaml = _interopRequireDefault(require("js-yaml"));

var _read = _interopRequireDefault(require("../read"));

var _write = _interopRequireDefault(require("../write"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable no-console */
async function main(argv) {
  const pkg = await (0, _cliBelt.loadPackage)(__dirname, {
    title: true
  });
  const help = _commonTags.stripIndent`
    ${pkg.description ? pkg.description : ''}

    Usage:
      $ yimp [options]

    Options:
      -i, --input <path>        Path to input file
      -o, --output <path>       Path to output file, optional
      -e, --ext <extensions>    Extensions, comma separated, optional
      -h, --help                Show help
      -v, --version             Show version number

    Example:
      $ yimp -i input-file.yml -o output-file.yml -e yml,yaml,raml
  `;
  const types = {
    '--input': String,
    '--output': String,
    '--ext': String,
    '--help': Boolean,
    '--version': Boolean
  };
  const {
    options: base,
    aliases
  } = (0, _cliBelt.flags)(help);
  (0, _cliBelt.safePairs)(types, base, {
    fail: true,
    bidirectional: true
  });
  Object.assign(types, aliases);
  const cmd = (0, _arg.default)(types, {
    argv,
    permissive: false,
    stopAtPositional: true
  });
  if (cmd['--help']) return console.log(help);
  if (cmd['--version']) return console.log(pkg.version);

  if (cmd._.length) {
    console.log(help + '\n');
    throw Error(`Unexpected command: ${cmd._[0]}`);
  }

  if (!cmd['--input']) {
    console.log(help + '\n');
    throw Error(`Input file path is required`);
  }

  const input = _path.default.isAbsolute(cmd['--input']) ? cmd['--input'] : _path.default.join(process.cwd(), cmd['--input']);
  const output = cmd['--output'] ? _path.default.isAbsolute(cmd['--output']) ? cmd['--output'] : _path.default.join(process.cwd(), cmd['--output']) : undefined;
  const options = cmd['--ext'] ? {
    ext: cmd['--ext'].split(',').map(ext => ext[0] === '.' ? ext : `.${ext}`)
  } : {};

  if (!output) {
    console.log(_jsYaml.default.dump((0, _read.default)(input, options)).trim());
  } else {
    (0, _write.default)(input, output, options);
  }
}
